### ISSUE TITLE
counting active sessions

### ISSUE THREAD

**Issue by bilias** [phpredis/phpredis#1878]
Hi,
I have a web application running on:
CentOS7
redis-3.2.12-2.el7.x86_64
php-pecl-redis5-5.3.2-1.el7.remi.7.3.x86_64
php-7.3.24-1.el7.remi.x86_64
fpm with php_value[session.gc_maxlifetime]  = 7200

I'm trying to count the active sessions (key ttl < 600 sec) using the code from https://github.com/phpredis/phpredis#scan
The problem is that I see spikes that are not normal. I see huge spikes (almost double sessions) at intervals of 90 minutes that are not real.

Any help/idea would be appreciated. I know that doing ttl() over all sessions is too much but I can't find another way to do it.
My code looks like this:

function getOnlineRedisUsers($redis) {
    $count = 0;
    $maxlifetime = ini_get('session.gc_maxlifetime');
    $max_ttl = $maxlifetime ? $maxlifetime : 7200;
    $max_ttl = 7200;
    $max_idle = 600;

    // https://github.com/phpredis/phpredis#scan
    $it = NULL;
    do {
        // Scan for some keys
        $arr_keys = $redis->scan($it, '', 10000);

        // Redis may return empty results, so protect against that
        if ($arr_keys !== FALSE) {
            foreach($arr_keys as $str_key) {
                $ttl = $redis->ttl($str_key);
                // -1: no ttl, -2: key does not exist
                if (($ttl > 0) and (($max_ttl-$ttl) < $max_idle)) {
                    $count++;
                }
            }
        }
    } while ($it > 0);

    return $count;
}

**Comment by bilias** [comment]
![sessions-day](https://user-images.githubusercontent.com/47315739/105849919-b7c39900-5fe9-11eb-9e25-f6116ef87810.png)

This is how it looks every day.

**Comment by michael-grunder** [comment]
Hi @bilias are you sure they're all session keys?  You could try limiting your scan to a pattern like `PHPREDIS_SESSION:*` or `PHPREDIS_CLUSTER_SESSION:*` if you're using Redis cluster.

Another alternative would be to `MONITOR` Redis when you expect the spike to occur to see if that shows anything useful.

**Comment by bilias** [comment]
Yes they are all "PHPREDIS_SESSION:

it is happening now, i don't see any strange things with monitor...

**Comment by bilias** [comment]
I also changed my code a little bit today to match https://github.com/phpredis/phpredis/issues/1372
without do {}
so the array is with unique keys. Somehow ttls are being messed up.

I'll try upgrading redis, cause I run the default from CentOS7 which is a bit old (3.2.12)

**Comment by michael-grunder** [comment]
I was just going to ask if it's possible you are counting some keys more than once :smile:

```php

function countUniqueSessions($redis, $prefix = 'PHPREDIS_SESSION:') {
    $arr_keys = [];

    $count = 0;
    $maxlifetime = ini_get('session.gc_maxlifetime');
    $max_ttl = $maxlifetime ? $maxlifetime : 7200;
    $max_ttl = 7200;
    $max_idle = 600;
    $arr_keys = [];

    $it = NULL;
    do {
        // Scan for some keys
        $arr_chunk = $redis->scan($it, $prefix . '*', 10000);
        if ($arr_chunk === false)
            continue;

        foreach ($arr_chunk as $str_key) {
            /* Skip duplicates */
            if (isset($arr_keys[$str_key]))
                continue;

            $ttl = $redis->ttl($str_key);
            if (($ttl > 0) and (($max_ttl-$ttl) < $max_idle)) {
                $arr_keys[$str_key] = $ttl;
            }
        }
    } while ($it > 0);

    return $arr_keys;
}
```

That script should avoid duplicate keys.

**Comment by bilias** [comment]
Yes as I told in my previous message, today I modified the code and did check for duplicates. That's not the problem :-/

btw (just for the record) your code need to change:
- function countUniqueSessions($redis, $prefix = 'PHPREDIS_SESSION:') {
+ function countUniqueSessions($redis, $prefix = 'PHPREDIS_SESSION:*') {

- return $arr_keys;
+ return count($arr_keys);

**Comment by michael-grunder** [comment]
I was returning the array because I'm also capturing each key's TTL.  I just thought it might be useful.

How do you know the extra session keys aren't real?  I'm not sure this will be solvable unless we figure out more information about the nature of these extra keys.

@yatsukhnenko Any ideas/questions?

**Comment by bilias** [comment]
No we don't have more keys in the DB.
Total remains the same.

ttl is getting malformed for some reason, at specific intervals as you can see from the graph.
No they are not real. We have equal time ranges if you see from the graph every day.

**Comment by yatsukhnenko** [comment]
@michael-grunder you are right, there's difficult to figure out what is wrong without PHP code that creates sessions. For example `session_regenerate_id` by default creates new session key and doesn't remove old one which may explain spikes at day time when users actively login on site.

**Comment by bilias** [comment]
But my total count() of keys in redis does not change when I have the spikes.

To my understanding, these are not new logins. It happens every day at 90 minutes intervals.
The login code is here, in case you want to have a look

https://github.com/gunet/openeclass/blob/7e36eb461aa2657cd8066471c9faadb413d95716/modules/auth/auth.inc.php#L1147

**Comment by yatsukhnenko** [comment]
> But my total count() of keys in redis does not change when I have the spikes.

Then what is shown at your cacti graph? How do you count sessions if it's not number of keys in redis?

**Comment by bilias** [comment]
I first get all keys with scan()
Then I individually check ttl() of each key. I need to count active users, say with idle time < 10min

**Comment by bilias** [comment]
I was wrong. Total sessions do increase at the same intervals.

@yatsukhnenko explanation with session_regenerate_id might be the one, and I see it's being used around the code.
Other relevant ideas?

thanks

**Comment by michael-grunder** [comment]
Going to close this since it's been inactive for a few years.

Hopefully you're not still having the issue!
